🐛 clusterctl only deletes the cluster object

juan-lee opened the pull request:
Signed-off-by: Juan-Lee Pang <jpang@microsoft.com>

**What this PR does / why we need it:**
Garbage collection will clean up any referenced MachineDeployment,
MachinMachineSet, and Machine objects.

**Which issue(s) this PR fixes** (optional, in fixes #<issue number>(, fixes #<issue_number>, ...) format, will close the issue(s) when PR gets merged):
Fixes #1444
Fixes #1469

ncdc commented on cmd/clusterctl/clusterdeployer/clusterdeployer_test.go:
```diff
@@ -284,6 +284,15 @@ func (c *testClusterClient) DeleteClusters(ns string) error {
 	} else {
 		delete(c.clusters, ns)
 	}
+	if err := c.DeleteMachineDeployments(ns); err != nil {
```
(Note, this isn't a comment about your changes, but one about the way this test was originally written 😄.) We should consider adjusting the test and its verification logic, because `TestClusterDelete` is really testing behavior of the `testClusterClient`. It seems off that your fix—remove the calls to delete MachineDeployments, MachineSets, and Machines—requires that you **add** these deletions to the `testClusterClient`. cc @chuckha

ncdc replied:
Or maybe a change you could make in this PR is to remove any expectation in the test code around MachineSets, MachineDeployments, and Machines. You could remove the checks for their deletions, and you could remove the test cases that involve errors creating and deleting these 3 resources.

juan-lee replied:
That sounds reasonable to me.

juan-lee replied:
I've removed the tests and validations that don't make sense anymore.

ncdc commented:
@juan-lee thanks for doing this, btw!

detiber commented on cmd/clusterctl/clusterdeployer/clusterdeployer.go:
```diff
@@ -216,29 +215,9 @@ func (d *ClusterDeployer) saveProviderComponentsToCluster(factory provider.Compo
-	}
-	if len(errorList) > 0 {
-		return errors.Errorf("error(s) encountered deleting objects from bootstrap cluster: [%v]", strings.Join(errorList, ", "))
+		return errors.Errorf("error encountered deleting objects from bootstrap cluster: [error deleting Clusters: %v]", err)
```
Will this run into any finalizer-related issues? I'm not overly familiar with the v1alpha2 related flow, but in v1alpha1 we removed the finalizers prior to delete to ensure that deletion wasn't blocked by the finalizer and that if the controllers were accidentally brought back up on the bootstrapping machine it wouldn't attempt to reconcile the resources that were already moved.

ncdc replied:
Oh, are we using this function to go in both directions (create: bootstrap -> workload, delete: workload -> bootstrap)?

vincepri replied:
We should probably block and wait for the cluster object to be deleted from the server, (unless client.DeleteClusters already does so). That should be enough to guarantee that the object tree has been deleted, in v1a2 the Cluster issues deletion requests to all of its children.

ncdc replied:
Looking at the code, I think this function is only used by `clusterctl delete`. Pivot uses different functions that do remove the finalizers (ForceDeleteXYZ)

ncdc replied:
@vincepri I'm pretty sure we added the same foreground deletion logic to v1alpha1, didn't we?

vincepri replied:
I think so, but only after a certain version

juan-lee replied:
DeleteAllOf was doing `metav1.DeletionPropagationOrphan` so I changed it to `metav1.DeletionPropagationForeground` so the object tree will be deleted and the call will block.

juan-lee replied:
When did that change go in? I did not retest after rebasing. I was previously sync'd to v0.2.3. In my testing the call to DeleteAllOf returned immediately. Looking at the code, the default behavior of the client is `metav1.DeletePropagationOrphan`.

ncdc replied:
Prior to releasing v0.2.0 (https://github.com/kubernetes-sigs/cluster-api/pull/1225).

Yes, it will return immediately - you'd have to add code to wait for it. When you run `kubectl delete cluster/foo`, it uses `DeletePropagatonBackground` and then waits for the cluster to be removed from the apiserver. It's because of kubectl's use of background propagation that we had to do what's in #1225.

juan-lee replied:
@ncdc PTAL at my approach to waiting for delete to complete.

wfernandes replied:
I took the approach by adding a method called `DeleteAllClusters` to the Cluster `Client` interface. I added a method that would just iterate through all the namespaces for clusters and delete each of them. 
https://github.com/wfernandes/cluster-api/blob/c644d0219dd970aebc27e0eb044c86481307b595/cmd/clusterctl/clusterdeployer/clusterclient/clusterclient.go#L540-L561

This could then be called by the [`clusterdeployer.go`](https://github.com/wfernandes/cluster-api/blob/c644d0219dd970aebc27e0eb044c86481307b595/cmd/clusterctl/clusterdeployer/clusterdeployer.go#L223)

This keeps the same behavior of the `Delete` method in `clusterdeployer.go` without having to call out to the `kubectl` binary. But I'm not strongly opposed to this method either. 
Apologies @juan-lee for my delay in creating my PR. I've been having a hard time creating AWS clusters using clusterctl and hence wasn't able to easily test my delete functionality.

wfernandes replied:
I also just updated the cluster functionality and not the other methods since we just have a `clusterctl delete cluster` command.

vincepri commented:
I changed the base branch to `release-0.2` on this PR

detiber commented:
> I changed the base branch to release-0.2 on this PR

Shouldn't this go against `master` and be backported to `release-0.2` in that case?

vincepri commented:
Yeah nevermind, changed it back

vincepri commented:
/test pull-cluster-api-test
--

wfernandes commented:
@juan-lee Are you able to get the intended behavior to work with `clusterctl`?

I've been working on #1469 and I noticed that the DeleteClusters operation fails for me. At first I  thought it was because of a finalizer issue but I noticed we weren't doing that in the `cluster_controller.go -> reconcileDelete`.

I notice that we are doing a `client.DeleteClusters("")` which provides an empty namespace. I believe this gets passed down all the way to [`typed_client.go->Delete()`](https://github.com/kubernetes-sigs/controller-runtime/blob/801e12a501606245062a6ff77ac91eef8936d9fb/pkg/client/typed_client.go#L72-L89).
Here it adds an empty namespace to the request since `clusterv1.Cluster` is Namespace'd scoped.

I believe we may have to delete the specific cluster object after determining the correct namespace.

ncdc commented:
@juan-lee @wfernandes do you all have any updates on the status of this (code changes and/or updated testing results)? Thanks in advance!

wfernandes commented:
@ncdc I was waiting to hear back from @juan-lee. Since this issue is closely related to the one I created (https://github.com/kubernetes-sigs/cluster-api/issues/1469) I didn't want to create any conflicts.
@juan-lee I'm happy to help in any way I can regarding manual testing, etc...

juan-lee commented:
Sorry for the late reply @wfernandes @ncdc. I'm trying to find time to finish this PR up today. @wfernandes if you're interested in helping me test if clusterctl is waiting for all of the owned objects to be deleted, I'd really appreciate it.

juan-lee commented:
@ncdc, I addressed the feedback in this PR. PTAL. I had to workaround issue #1469 that @wfernandes  has been working on to test the deletion propagation behavior. I did not address that bug in this PR since I believe that @wfernandes will provide a fix separately.

juan-lee commented:
/retest

vincepri commented:
/milestone v0.3.0

ncdc commented on cmd/clusterctl/clusterdeployer/clusterclient/clusterclient.go:
```diff
@@ -809,21 +801,45 @@ func (c *client) kubectlManifestCmd(commandName, manifest string) error {
+	args := []string{"delete", resource}
+	if namespace != "" {
+		args = append(args, "--namespace", namespace)
+	}
```
If you want to use kubectl to delete from all namespaces, you'll need to specify `--all-namespaces`; otherwise, it will only delete from the namespace for the kubeconfig's context.

juan-lee replied:
good catch @ncdc

ncdc commented on cmd/clusterctl/clusterdeployer/clusterclient/clusterclient.go:
```diff
@@ -795,12 +783,16 @@ func (c *client) kubectlDelete(manifest string) error {
 	return c.kubectlManifestCmd("delete", manifest)
 }
 
+func (c *client) kubectlDeleteAll(namespace, resource string) error {
```
Do we need this function? It looks like it's a 1:1 wrap around `c.kubectlDeleteAllCmd`

ncdc commented:
@juan-lee have you tested that create / pivot / delete all continue to work as before?

juan-lee commented:
> @juan-lee have you tested that create / pivot / delete all continue to work as before?

I have tested create / pivot / delete and they all work now. As discussed earlier in this PR, both @wfernandes and I had issues getting delete to work prior to these changes.

ncdc commented:
@detiber / @vincepri - I'll defer to y'all since you have more experience w/clusterctl code

juan-lee commented:
> @detiber / @vincepri - I'll defer to y'all since you have more experience w/clusterctl code

The reason the delete all clusters functionality was broken was because we were passing "" for namespace to DeleteAllOf. At some point passing "" must have changed from selecting all namespace to selecting no namespaces.

wfernandes commented:
Fixes https://github.com/kubernetes-sigs/cluster-api/issues/1469

Adding this comment so this issue gets closed if we decide to merge this PR. I'm assuming that the prow bot will figure out this directive.

wfernandes commented:
https://github.com/kubernetes-sigs/cluster-api/pull/1445#discussion_r337456205

Adding the above link for context in the main thread since I realized I put it on a review thread above. Apologies for any confusion.

ncdc commented:
@wfernandes unfortunately only `fixes` directives in the PR's description (1st "comment") are processed (by GitHub, not prow). I went ahead and edited the description to include it for you.

detiber commented:
No objections from me, @vincepri

vincepri commented:
/assign
